File: packages/ingest/src/lib/mgl/adapter.ts
import type { NodeMeta } from "../../types";
import type {
	GenericWorkflowAdapter,
	RootPlan,
	ShardWorkItem,
	UnitPlan,
} from "../ingest/adapter-types";
import { extractSectionCrossReferences } from "./cross-references";
import {
	fetchMglChapter,
	fetchMglLandingHtml,
	fetchMglPart,
	fetchMglParts,
	fetchMglSection,
} from "./fetcher";
import {
	designatorSortOrder,
	extractVersionIdFromLandingHtml,
	type MglPart,
	parseChapterDetail,
	parsePartDetail,
	parsePartSummary,
	parseSectionContent,
	parseSectionSummary,
} from "./parser";
import { normalizeMglApiUrl, normalizeMglPublicUrl } from "./utils";

const SOURCE_CODE = "mgl";
const SOURCE_NAME = "Massachusetts General Laws";
const SECTION_LEVEL_INDEX = 2;
const BLOB_STORE_BATCH_SIZE = 50;

interface MglUnitRoot extends MglPart {
	id: string;
}

type MglShardMeta =
	| { kind: "node"; node: NodeMeta }
	| {
			kind: "section";
			partCode: string;
			chapterCode: string;
			sectionCode: string;
			sectionApiUrl: string;
			sortOrder: number;
	  };

type MglSectionShardItem = ShardWorkItem<
	Extract<MglShardMeta, { kind: "section" }>
>;

function createRootNodePlan(versionId: string, startUrl: string): NodeMeta {
	return {
		id: `${SOURCE_CODE}/${versionId}/root`,
		source_version_id: "",
		parent_id: null,
		level_name: "root",
		level_index: -1,
		sort_order: 0,
		name: SOURCE_NAME,
		path: "/statutes/mgl",
		readable_id: "MGL",
		heading_citation: "MGL",
		source_url: startUrl,
		accessed_at: new Date().toISOString(),
	};
}

function makePartId(rootId: string, partCode: string): string {
	return `${rootId}/part-${partCode.toLowerCase()}`;
}

function makeChapterId(partId: string, chapterCode: string): string {
	return `${partId}/chapter-${chapterCode.toLowerCase()}`;
}

function makeSectionId(chapterId: string, sectionCode: string): string {
	return `${chapterId}/section-${sectionCode.toLowerCase()}`;
}

function normalizeName(name: string): string {
	return name.replace(/\s+/g, " ").trim();
}

function getApiUrl(rawUrl: string, baseUrl: string): string {
	return (
		normalizeMglApiUrl(rawUrl, baseUrl) ?? new URL(rawUrl, baseUrl).toString()
	);
}

function chapterPageUrl(baseUrl: string, chapterCode: string): string {
	return `${baseUrl}/laws/generallaws/chapter${chapterCode.toLowerCase()}`;
}

export const mglAdapter: GenericWorkflowAdapter<MglUnitRoot, MglShardMeta> = {
	source: {
		code: SOURCE_CODE,
		name: SOURCE_NAME,
		jurisdiction: "state",
		region: "MA",
		docType: "statute",
	},
	maxUnitConcurrency: 1,

	async discoverRoot({ env }): Promise<RootPlan<MglUnitRoot>> {
		const startUrl = `${env.MGL_BASE_URL}${env.MGL_START_PATH}`;
		const rootHtml = await fetchMglLandingHtml(startUrl);
		const versionId = extractVersionIdFromLandingHtml(rootHtml);

		await env.STORAGE.put(`sources/mgl/${versionId}/root.html`, rootHtml, {
			httpMetadata: { contentType: "text/html" },
		});

		const partSummaries = await fetchMglParts(env, versionId, env.MGL_BASE_URL);
		const units: MglUnitRoot[] = [];
		for (const summary of partSummaries) {
			const summaryPart = parsePartSummary(
				summary,
				getApiUrl(summary.Details, env.MGL_BASE_URL),
			);
			const detail = await fetchMglPart(
				env,
				versionId,
				env.MGL_BASE_URL,
				summaryPart.partCode,
			);
			const parsed = parsePartDetail(
				detail,
				getApiUrl(summary.Details, env.MGL_BASE_URL),
			);
			units.push({
				...parsed,
				id: `part-${parsed.partCode.toLowerCase()}`,
			});
		}

		units.sort((a, b) => a.sortOrder - b.sortOrder);

		return {
			versionId,
			rootNode: createRootNodePlan(versionId, startUrl),
			unitRoots: units,
		};
	},

	async planUnit({ env, root, unit }): Promise<UnitPlan<MglShardMeta>> {
		const accessedAt = new Date().toISOString();
		const partId = makePartId(root.rootNode.id, unit.partCode);
		const shardItems: Array<ShardWorkItem<MglShardMeta>> = [];
		const seenSectionIds = new Set<string>();

		const partNode: NodeMeta = {
			id: partId,
			source_version_id: root.sourceVersionId,
			parent_id: root.rootNode.id,
			level_name: "part",
			level_index: 0,
			sort_order: unit.sortOrder,
			name: normalizeName(unit.partName),
			path: `/statutes/mgl/part/${unit.partCode.toLowerCase()}`,
			readable_id: `Part ${unit.partCode}`,
			heading_citation: `Part ${unit.partCode}`,
			source_url: unit.partApiUrl,
			accessed_at: accessedAt,
		};

		shardItems.push({
			parentId: root.rootNode.id,
			childId: partId,
			sourceUrl: unit.partApiUrl,
			meta: { kind: "node", node: partNode },
		});

		const part = await fetchMglPart(
			env,
			root.versionId,
			env.MGL_BASE_URL,
			unit.partCode,
		);
		const chapterSummaries = [...part.Chapters].sort(
			(a, b) => designatorSortOrder(a.Code) - designatorSortOrder(b.Code),
		);

		for (const chapterSummary of chapterSummaries) {
			const chapterApiUrl = getApiUrl(chapterSummary.Details, env.MGL_BASE_URL);
			const chapterDetail = await fetchMglChapter(
				env,
				root.versionId,
				env.MGL_BASE_URL,
				chapterSummary.Code,
			);
			const chapter = parseChapterDetail(chapterDetail, chapterApiUrl);
			const chapterId = makeChapterId(partId, chapter.chapterCode);
			const publicChapterUrl =
				normalizeMglPublicUrl(chapterSummary.Details, env.MGL_BASE_URL) ??
				chapterPageUrl(env.MGL_BASE_URL, chapter.chapterCode);

			const chapterNode: NodeMeta = {
				id: chapterId,
				source_version_id: root.sourceVersionId,
				parent_id: partId,
				level_name: "chapter",
				level_index: 1,
				sort_order: chapter.sortOrder,
				name: normalizeName(chapter.chapterName),
				path: `/statutes/mgl/part/${unit.partCode.toLowerCase()}/chapter/${chapter.chapterCode.toLowerCase()}`,
				readable_id: `Chapter ${chapter.chapterCode}`,
				heading_citation: `Chapter ${chapter.chapterCode}`,
				source_url: publicChapterUrl,
				accessed_at: accessedAt,
			};

			shardItems.push({
				parentId: partId,
				childId: chapterId,
				sourceUrl: chapter.chapterApiUrl,
				meta: { kind: "node", node: chapterNode },
			});

			const sectionSummaries = [...chapterDetail.Sections].sort(
				(a, b) => designatorSortOrder(a.Code) - designatorSortOrder(b.Code),
			);
			for (const sectionSummary of sectionSummaries) {
				const section = parseSectionSummary(
					sectionSummary,
					getApiUrl(sectionSummary.Details, env.MGL_BASE_URL),
				);
				const sectionId = makeSectionId(chapterId, section.sectionCode);
				if (seenSectionIds.has(sectionId)) continue;
				seenSectionIds.add(sectionId);

				shardItems.push({
					parentId: chapterId,
					childId: sectionId,
					sourceUrl: section.sectionApiUrl,
					meta: {
						kind: "section",
						partCode: unit.partCode,
						chapterCode: chapter.chapterCode,
						sectionCode: section.sectionCode,
						sectionApiUrl: section.sectionApiUrl,
						sortOrder: section.sortOrder,
					},
				});
			}
		}

		return {
			unitId: unit.id,
			shardItems,
		};
	},

	async loadShardItems({
		env,
		root,
		sourceVersionId,
		items,
		nodeStore,
		blobStore,
	}): Promise<void> {
		const accessedAt = new Date().toISOString();
		const sectionItems: MglSectionShardItem[] = [];
		const pendingWrites: Array<{ node: NodeMeta; content: unknown }> = [];

		for (const item of items) {
			if (item.meta.kind === "node") {
				await nodeStore.store(item.meta.node, null);
				continue;
			}
			sectionItems.push(item as MglSectionShardItem);
		}

		if (sectionItems.length === 0) return;

		const flushPendingWrites = async () => {
			if (pendingWrites.length === 0) return;
			const blobHashes = await blobStore.storeJsonBatch(
				pendingWrites.map((entry) => entry.content),
			);
			for (let i = 0; i < pendingWrites.length; i++) {
				await nodeStore.store(pendingWrites[i].node, blobHashes[i]);
			}
			pendingWrites.length = 0;
		};

		for (const item of sectionItems) {
			const section = await fetchMglSection(
				env,
				root.versionId,
				env.MGL_BASE_URL,
				item.meta.chapterCode,
				item.meta.sectionCode,
			);
			const parsed = parseSectionContent(section);
			const crossReferences = extractSectionCrossReferences(parsed.body);
			const readableId = `MGL c.${item.meta.chapterCode} §${item.meta.sectionCode}`;
			const path = `/statutes/mgl/part/${item.meta.partCode.toLowerCase()}/chapter/${item.meta.chapterCode.toLowerCase()}/section/${item.meta.sectionCode.toLowerCase()}`;

			pendingWrites.push({
				node: {
					id: item.childId,
					source_version_id: sourceVersionId,
					parent_id: item.parentId,
					level_name: "section",
					level_index: SECTION_LEVEL_INDEX,
					sort_order: item.meta.sortOrder,
					name: parsed.heading,
					path,
					readable_id: readableId,
					heading_citation: readableId,
					source_url: item.meta.sectionApiUrl,
					accessed_at: accessedAt,
				},
				content: {
					blocks: [{ type: "body", content: parsed.body }],
					...(crossReferences.length > 0
						? { metadata: { cross_references: crossReferences } }
						: {}),
				},
			});

			if (pendingWrites.length >= BLOB_STORE_BATCH_SIZE) {
				await flushPendingWrites();
			}
		}

		await flushPendingWrites();
	},
};

export type { MglShardMeta, MglUnitRoot };

File: packages/ingest/container-rust/src/sources/mgl/adapter.rs
use crate::runtime::fetcher::Fetcher;
use crate::runtime::types::{IngestContext, QueueItem};
use crate::sources::common::body_block;
use crate::sources::mgl::cross_references::inline_section_cross_references;
use crate::sources::mgl::parser::{
    designator_sort_order, normalize_body_text, normalize_designator, parse_chapter_detail,
    parse_part_detail, MglApiChapter, MglApiPart, MglApiSection,
};
use crate::sources::SourceAdapter;
use crate::types::{DiscoveryResult, NodeMeta, NodePayload, SectionContent};
use async_trait::async_trait;
use serde_json::json;

pub struct MglAdapter;

pub const MGL_ADAPTER: MglAdapter = MglAdapter;

#[async_trait]
impl SourceAdapter for MglAdapter {
    async fn discover(&self, fetcher: &dyn Fetcher, url: &str) -> Result<DiscoveryResult, String> {
        crate::sources::mgl::discover::discover_mgl_root(fetcher, url).await
    }

    async fn process_url(
        &self,
        context: &mut IngestContext<'_>,
        item: &QueueItem,
    ) -> Result<(), String> {
        let url = &item.url;
        let metadata = &item.metadata;
        match item.level_name.as_str() {
            "unit" | "part" => {
                let title_num = metadata["title_num"].as_str().unwrap_or_default();

                let json_str = context
                    .cache
                    .fetch_cached(url, &format!("mgl/part-{}.json", title_num))
                    .await?;
                let part: MglApiPart = serde_json::from_str(&json_str)
                    .map_err(|err| format!("Failed to parse MGL part JSON: {url}: {err}"))?;

                let parsed_part = parse_part_detail(&part, url);

                // Emit part node
                let part_id = format!(
                    "{}/part-{}",
                    context.build.root_node_id,
                    title_num.to_lowercase()
                );
                context
                    .nodes
                    .insert_node(NodePayload {
                        meta: NodeMeta {
                            id: part_id.clone(),
                            source_version_id: context.build.source_version_id.to_string(),
                            parent_id: Some(context.build.root_node_id.to_string()),
                            level_name: "part".to_string(),
                            level_index: 0,
                            sort_order: parsed_part.sort_order,
                            name: Some(parsed_part.part_name.clone()),
                            path: Some(format!("/statutes/mgl/part/{}", title_num.to_lowercase())),
                            readable_id: Some(title_num.to_string()),
                            heading_citation: Some(format!("Part {}", title_num)),
                            source_url: Some(url.to_string()),
                            accessed_at: Some(context.build.accessed_at.to_string()),
                        },
                        content: None,
                    })
                    .await?;

                // Enqueue chapters
                for chapter_summary in &part.Chapters {
                    let chapter_url = chapter_summary.Details.replace("http://", "https://");
                    context.queue.enqueue(QueueItem {
                        url: chapter_url,
                        parent_id: part_id.clone(),
                        level_name: "chapter".to_string(),
                        level_index: 1,
                        metadata: json!({
                            "title_num": title_num,
                            "chapter_code": chapter_summary.Code
                        }),
                    });
                }
            }
            "chapter" => {
                let title_num = metadata["title_num"].as_str().unwrap_or_default();
                let chapter_code = metadata["chapter_code"].as_str().unwrap_or_default();

                let cache_key = format!("mgl/chapter-{}.json", chapter_code.to_lowercase());
                let json_str = context.cache.fetch_cached(url, &cache_key).await?;
                let chapter: MglApiChapter = serde_json::from_str(&json_str)
                    .map_err(|err| format!("Failed to parse MGL chapter JSON: {url}: {err}"))?;

                let parsed_chapter = parse_chapter_detail(&chapter, url);

                // Emit chapter node
                let chapter_id = format!(
                    "{}/chapter-{}",
                    item.parent_id,
                    parsed_chapter.chapter_code.to_lowercase()
                );
                context
                    .nodes
                    .insert_node(NodePayload {
                        meta: NodeMeta {
                            id: chapter_id.clone(),
                            source_version_id: context.build.source_version_id.to_string(),
                            parent_id: Some(item.parent_id.clone()),
                            level_name: "chapter".to_string(),
                            level_index: 1,
                            sort_order: parsed_chapter.sort_order,
                            name: Some(parsed_chapter.chapter_name.clone()),
                            path: Some(format!(
                                "/statutes/mgl/part/{}/chapter/{}",
                                title_num.to_lowercase(),
                                parsed_chapter.chapter_code.to_lowercase()
                            )),
                            readable_id: Some(parsed_chapter.chapter_code.clone()),
                            heading_citation: Some(format!(
                                "Chapter {}",
                                parsed_chapter.chapter_code
                            )),
                            source_url: Some(url.to_string()),
                            accessed_at: Some(context.build.accessed_at.to_string()),
                        },
                        content: None,
                    })
                    .await?;

                // Enqueue sections
                let mut sections = chapter.Sections.clone();
                sections.sort_by_key(|s| designator_sort_order(&s.Code));

                for (i, section_data) in sections.into_iter().enumerate() {
                    let section_code = normalize_designator(&section_data.Code);
                    let section_url = section_data
                        .Details
                        .as_deref()
                        .unwrap_or(url)
                        .replace("http://", "https://");

                    context.queue.enqueue(QueueItem {
                        url: section_url,
                        parent_id: chapter_id.clone(),
                        level_name: "section".to_string(),
                        level_index: 2,
                        metadata: json!({
                            "title_num": title_num,
                            "chapter_code": parsed_chapter.chapter_code,
                            "section_code": section_code,
                            "sort_order": i as i32,
                            "immediate_text": section_data.Text,
                            "immediate_name": section_data.Name
                        }),
                    });
                }
            }
            "section" => {
                let title_num = metadata["title_num"].as_str().unwrap_or_default();
                let chapter_code = metadata["chapter_code"].as_str().unwrap_or_default();
                let section_code = metadata["section_code"].as_str().unwrap_or_default();
                let sort_order = metadata["sort_order"].as_i64().unwrap_or(0) as i32;

                let mut raw_body = metadata["immediate_text"]
                    .as_str()
                    .unwrap_or_default()
                    .to_string();
                let mut section_name_opt =
                    metadata["immediate_name"].as_str().map(|s| s.to_string());

                if raw_body.trim().is_empty() && url != "none" {
                    let cache_key = format!(
                        "mgl/chapter-{}-section-{}.json",
                        chapter_code.to_lowercase(),
                        section_code.to_lowercase()
                    );
                    match context.cache.fetch_cached(url, &cache_key).await {
                        Ok(json_str) => {
                            if let Ok(full_section) =
                                serde_json::from_str::<MglApiSection>(&json_str)
                            {
                                if let Some(text) = full_section.Text {
                                    raw_body = text;
                                }
                                if let Some(name) = full_section.Name {
                                    section_name_opt = Some(name);
                                }
                            }
                        }
                        Err(e) => {
                            eprintln!("Failed to fetch section details for {section_code}: {e}");
                        }
                    }
                }

                let normalized = normalize_body_text(&raw_body);
                let body = inline_section_cross_references(&normalized);

                let blocks = vec![body_block(&body)];

                let content = SectionContent {
                    blocks,
                    metadata: None,
                };
                let section_id =
                    format!("{}/section-{}", item.parent_id, section_code.to_lowercase());
                let heading_citation = format!("MGL c.{} §{}", chapter_code, section_code);
                let section_name = section_name_opt.unwrap_or_else(|| section_code.to_string());

                context
                    .nodes
                    .insert_node(NodePayload {
                        meta: NodeMeta {
                            id: section_id,
                            source_version_id: context.build.source_version_id.to_string(),
                            parent_id: Some(item.parent_id.clone()),
                            level_name: "section".to_string(),
                            level_index: 2,
                            sort_order,
                            name: Some(section_name),
                            path: Some(format!(
                                "/statutes/mgl/part/{}/chapter/{}/section/{}",
                                title_num.to_lowercase(),
                                chapter_code.to_lowercase(),
                                section_code.to_lowercase()
                            )),
                            readable_id: Some(section_code.to_string()),
                            heading_citation: Some(heading_citation),
                            source_url: Some(url.to_string()),
                            accessed_at: Some(context.build.accessed_at.to_string()),
                        },
                        content: Some(serde_json::to_value(&content).unwrap()),
                    })
                    .await?;
            }
            other => return Err(format!("Unknown MGL level: {other}")),
        }

        Ok(())
    }

    fn unit_label(&self, item: &QueueItem) -> String {
        match item.level_name.as_str() {
            "unit" | "part" => format!(
                "Part {}",
                item.metadata["title_num"].as_str().unwrap_or("?")
            ),
            "chapter" => format!(
                "Chapter {}",
                item.metadata["chapter_code"].as_str().unwrap_or("?")
            ),
            "section" => format!(
                "Section {}",
                item.metadata["section_code"].as_str().unwrap_or("?")
            ),
            other => other.to_string(),
        }
    }
    fn needs_zip_extraction(&self) -> bool {
        false
    }
}
